### PR TITLE
Rework maven poms, dependencies, and version management

### LINKED ISSUE #2: Deployment to Maven Central
Could you deploy the artifacts to Maven Central? Then we could easily integrate them.

### LINKED ISSUE #26: Fix SLF4J warning about missing logger
The following warning is generated each time the tests are run:
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
```

I think this can be fixed by adding this to the pom.xml:
```xml
       <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
            <version>2.0.5</version>
       </dependency>
       <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-simple</artifactId>
            <version>2.0.5</version>
       </dependency>
```
(I don't know if the simple one is the right logger)

### PR DESCRIPTION
**Pull request by dfuchss**
Changes in this PR:

1. simplified the dependencies in the poms, removed redundancies and added consistent versions.
2. Added Dependabot configuration for automatic updates
3. set the versions of several maven related plugins
4. reworked the bundling mechanism of the CLI in order to get simple libraries for each module (without shade plugin)
5. added more actions to the CI
6. fixed the logger config 
7. fixed the execution of dynamic tests (some renaming / refactoring had been missing)
8. fixed the loading of the included agent/executor jar

Missing: 
- [x] Check that deployment works (e.g., with snapshot snapshot repository; currently parent is missing in the current release from the main branch, that's why I worked on it.)
- [x] If everything works as expected, I'll update the docs

Resolves #26 
Resolves #2

### CODE REVIEW

**Comment by dfuchss** [comment]
@Feuermagier if you find some time, please have a look at the changes :)

**Comment by dfuchss** [comment]
@Feuermagier I'm not sure whether the `ClassTransformer` has to be modified:

should the transformer transform also the autograder? Otherwise, I guess `className.startsWith("de/firemage/autograder")` is missing

**Comment by dfuchss** [comment]
@Feuermagier one last question: Why is ENABLE_DYNAMIC set to false (in the CheckTest) ?

File: .github/dependabot.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+version: 2
```
Comment (dfuchss) [inline .github/dependabot.yml:1]: This will generate PRs for github actions and maven dependencies every week

File: autograder-agent/pom.xml [hunk 2]
```diff
@@ -1,83 +1,67 @@
...
 <project xmlns="http://maven.apache.org/POM/4.0.0"
-         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <parent>
         <artifactId>autograder-parent</artifactId>
         <groupId>de.firemage.autograder</groupId>
-        <version>0.1</version>
+        <version>${revision}</version>
```
Comment (dfuchss) [inline autograder-agent/pom.xml:8]: Revision is set using the properties in the parent pom.xml

File: autograder-core/src/main/java/de/firemage/autograder/core/integrated/IntegratedAnalysis.java [hunk 3]
```diff
@@ -39,16 +52,50 @@ public IntegratedAnalysis(UploadedFile file, Path jar, Path tmpPath, Consumer<Li
...
+        if (resource == null) {
+            throw new IllegalArgumentException("URL is null");
+        }
+        URI uri = resource.toURI();
+        if (!uri.toString().contains("!")) {
+            return Path.of(uri);
+        }
+        // See https://stackoverflow.com/questions/22605666/java-access-files-in-jar-causes-java-nio-file-filesystemnotfoundexception
+        String[] path = uri.toString().split("!", 2);
+        @SuppressWarnings("resource") FileSystem fs = createFileSystem(path[0]);
```
Comment (dfuchss) [inline autograder-core/src/main/java/de/firemage/autograder/core/integrated/IntegratedAnalysis.java:75]: https://github.com/Feuermagier/autograder/pull/25/files#diff-624cfed3224cd3668e334c207219d10e516920ff3039e75faf40c72b1ce8715dR90 handles the close operation of `fs`

File: autograder-executor/pom.xml [hunk 4]
```diff
@@ -5,80 +5,62 @@
...
                 </configuration>
                 <executions>
                     <execution>
-                        <phase>package</phase>
+                        <id>make-assembly</id>
                         <goals>
-                            <goal>shade</goal>
+                            <goal>single</goal>
                         </goals>
+                        <phase>compile</phase>
```
Comment (dfuchss) [inline autograder-executor/pom.xml:60]: compile generates a jar-with-dependencies that will be copied during compile (into the core project)

File: pom.xml [hunk 5]
```diff
@@ -131,6 +252,26 @@
...
             </build>
+            <modules>
+                <module>autograder-core</module>
+                <module>autograder-executor</module>
+                <module>autograder-event</module>
+                <module>autograder-agent</module>
+            </modules>
+        </profile>
+
+        <profile>
```
Comment (dfuchss) [inline pom.xml:263]: this profile is active if release profile is not active

**Review by dfuchss — COMMENTED** [review]
Some final comments.

**Comment by Feuermagier** [comment]
Thanks for cleaning up the mess I made with Maven :D

Dynamic analysis is currently disabled in checks mostly because of execution time. It's also disabled by the Grading Tool, as I don't think everyone would be comfortable with executing the student's code on their local machine, even if it is inside a resource-constrained Docker container.

**Comment by Feuermagier** [comment]
Deployment works: https://s01.oss.sonatype.org/content/repositories/snapshots/de/firemage/autograder/

Do you want to do any more changes on this branch? If not, I'll merge.

**Comment by dfuchss** [comment]
You can merge it, if everything is fine from your side.

**Pull request merged**